Add particle binning to picmi

chillenzer opened the pull request:
## Particle Binning for PICMI

This PR adds initial support for particle binning to PICMI. This support is not yet complete, see list below for what's missing. The interface is to be considered experimental and subject to change (because it is yet unclear if the interface will have to distinguish between particle and field binning explicitly). Because of that there's no documentation at the moment. Power-users like @pordyna will figure it out and others can look at the description in this PR. Experienced users should check the generated C++ code if things don't do what's expected.

### Status of this PR and your mission

Please review, test and start using this PR, so we can get decent feedback about this. ~~It's not completely polished yet as it would benefit from some rebasing and working on this has uncovered some bugs that probably have to be fixed in C++ world. I'll leave this as green, so people will actually care to look at it. But I expect no review to pass in every detail yet.~~


### UI Description

The UI can be seen in the end-to-end test. Users can add an arbitrary number of `picmi.diagnostics.binning.Binning` objects to `sim.diagnostics` each of which contains four necessary pieces of information:
```python
            Binning(
                name="particle_density",
                deposition_functor=BinningFunctor(
                    name="particle_density",
                    functor=lambda particle: particle.get("weighting"),
                    # Can be arithmetic python types or a string which would be inserted literally:
                    return_type=float,
                ),
                axes=[BinningAxis(...), ...],
                # could be a list:
                species=species_hydrogen,
                # and more with defaults...
                period=TimeStepSpec[:],
                # ...most of the other things that the `Binner` has as setters...
            )
```
This will produce the desired binning output in the folder `${run_dir}/simOutput/binningOpenPMD/${Binning.name}_${infix}.${extension}`. Some details:
- The `BinningFunctor` takes a function with the signature `def functor(particle)`. The particle passed to the function exposes a `.get` method that can take any particle attribute name as string (anything that we'd get via `particle[attribute_]` in C++) as well as `position`, `timestep` and the attributes implemented as traits like `picongpu::traits::attribute::getMass`. For `position` it can take additional keyword arguments `origin`, `precision` and `unit` (but see below because that's not yet fully functional). The return value of `particle.get` is either a `sympy.Symbol` or a tuple of `sympy.Symbol` (depending on whether a scalar or a vector was asked for). We might find a more appropriate collection later on. The obvious candidates `sympy.Vector` and/or `sympy.Point` unfortunately have their own quirks and bugs and also a relatively bulky interface, so I'm not yet sure about what's the best interface here.
- The `BinningAxis` class bundles a `BinningFunctor` and a `BinSpec(kind="linear"/"log", start, stop, nsteps)`.
- The further configuration options include all the simple things like `timeaveraging`, openPMD configuration (as a dictionary that's converted to a json string for code generation) and so on. It does not contain the tricky stuff like host-side hooks and openPMD functors. These will likely never get exposed but will only be used internally as means to provide higher level features in the frontend.


### Implementation details

The translation of functors from the UI to the generated code is roughly organised as follows:
- On the PICMI level, the interface is provided. Upon translation into the PyPIConGPU layer via `.get_as_pypicongpu`, the `BinningFunctor` calls the given functor with an opaque `BinningParticle` object. While it returns the corresponding symbols from the `.get` method, it keeps track of the requested properties and handed out symbols. The sympy expression as well as a dictionary mapping symbols to attributes is passed to PyPIConGPU.
- On the PyPIConGPU level, the `functor_expression` is handled as in the free-formula density case via the `PMAccPrinter`. This produces an expression containing the variables as they were handed out as `sympy.Symbol`s above. In order to declare and define these variables, PyPIConGPU's `BinningFunctor` generates a custom preamble to preceed the generated `functor_expression` in the generated code. It roughly looks like this (wherein, of course, only statements for actually used variables are generated)
```C++
auto const weighting = particle[weighting_];
auto const [px, py, pz] = particle[momentum_];
// but also things like...
auto const [x_cell_cell,y_cell_cell,z_cell_cell] = getParticlePosition<DomainOrigin::TOTAL, PositionPrecision::Cell, PositionUnits::Cell>(domainInfo, particle);
```
- The template organises each `Binner` into its own function and calls all those functions in `getBinning` at the end.

The other data are straightforwardly passed through the layers and not much magic is happening. I have also restrained from putting in strong guardrails at the moment because a number of aspects are planned as (or at least conceivable as) extension points or even explicitly allow for arbitrary strings to be passed through.


### Testing
Tests for this feature are built on the infrastructure set up for free-formula densities and are expected to suffer from the same problems in CI, so there's no point in activating them at the moment. This makes it even more important that the reviewers please run the tests! I will come up with a better separation of concerns in the process of bringing the tests into CI.

The main test again computes the particle density and compares to the other methods that already existed. I'm planning to add some further tests for different functionalities but success in this direction was limited by the backend bugs described below. As submitted, the tests currently do not succeed but fail at compile time. Deactivating the (as of now unused-in-the-tests) position binnings makes them pass.


### Missing features
In general, this PR does not yet attempt to add field binning. Furthermore, not all details from the particle binning are implemented as of now. This is the list of missing feature I'm aware of. Please ping me and add to this list if there's anything I've overlooked.
- Add units to binning.
- Species filtering.
~~- Add supercell as precision and potentially local cell index as origin, etc.~~
- Accumulate operation is not configurable.


~~Backend Bugs~~

~~Flexible binning is a relatively young feature by itself, so there's still some bugs in the backend that have been triggered by the systematic testing for this PR:~~

~~- Some initial tests ran into segfaults with more than 5 axes. @ikbuibui is supposed to investigate this further.~~
~~- The position functors which are currently added but not yet tested lead to compile-time errors about inconvertible return types. @ikbuibui is supposed to investigate this further.~~

ikbuibui commented:
I looked into the issues you faced. 
1. I tried your setup and indeed ran into a segfault if I enable that axis. This happens not because of a limit on supported axes but because we try to allocate a 64 GiB buffer on host and device and we run out of memory. I tried decreasing the size of the binning axes and was able to test a binning of 8 axes. 
2. The errors regarding inconvertible return types occur because the "backend" currently doesn't support integral cell sizes. I have added the fix for this to #5402. At the heart of this lies a deeper issue regarding lack of conversions in arithmetic operators of `pmacc::Vector`, which we dont do to avoid silent-ish upcasting of types inadvertently by the user.

ikbuibui commented:
It would make sense to add access to ``picongpu/algorithms/``, ``picongpu/traits/attribute`` and some of the symbols available from ``simulation.unitless`` like `getEps0` and `getSpeedOfLight` in the binning functors as these might be useful for the physicists

chillenzer committed:
Fix missing LOCAL_WITH_GUARDS in ValidPositionRequest

chillenzer committed:
Add rudimentary particle binning to pypicongpu

chillenzer committed:
Add rudimentary particle binning to picmi

chillenzer committed:
Add binning end-to-end tests and refactor

chillenzer commented:
This PR is ready for review. Notable changes since last push:
- Included a fix to `ValidPositionRequest` that adds `LOCAL_WITH_GUARDS`.
- Completion of origin, precision and unit support for domain info requests.
- Added `.get` access to pertinent functionality from `picongpu/algorithms`
- Refactored the end-to-end tests for better readability and structure. (They are still not included in CI and still need a general overhaul but it's a step forward I hope.)
- Cleaned up the history into a digestible number of commits.
- Potentially, there are futher minor things.

ikbuibui commented on lib/python/picongpu/picmi/diagnostics/binning.py:
```diff
@@ -0,0 +1,191 @@
+            self.get("momentum")
+            if attribute == "gamma":
+                symbols = sympy.Symbol("gamma")
+            if attribute == "kinetic energy":
```
```suggestion
            elif attribute == "kinetic energy":
```
Minor thing, but logically `elif` seems more appropriate. Same thing also happens below with velocity.

ikbuibui commented on lib/python/picongpu/pypicongpu/output/binning.py:
```diff
@@ -0,0 +1,177 @@
+} | {
+    "mass": "picongpu::traits::attribute::getMass(particle[weighting_], particle)",
+    # CAUTION: The names in the gamma formula are currently hardcoded.
+    # We'll certainly trip over this, should we ever dare to change the internal names.
```
🙃

ikbuibui reviewed (commented):
Aside from the fact that I need to create a directory called `scratch` in my home to hold the `end-to-end` test output, things look okay.  Tests pass on my machine.
The refactoring of the distributions and the related test is sneaked in but its fine.

BrianMarre commented on lib/python/picongpu/picmi/diagnostics/binning.py:
```diff
@@ -0,0 +1,191 @@
+        name: str,
+        deposition_functor: BinningFunctor,
+        axes: list[BinningAxis],
+        species: PICMISpecies | list[PICMISpecies],
```
I would simplify the interface here and always require a list of species

chillenzer replied:
Nope, I don't want that. That's precisely the kind of thing that leads to users coming over and complaining why they have to cast every numpy array to a list. The intent of the user is obvious when they give a single species, so we should use the one obvious interpretation and go with it.

BrianMarre commented on lib/python/picongpu/picmi/diagnostics/binning.py:
```diff
@@ -0,0 +1,191 @@
+        openPMD: Optional[dict] = None,
+        openPMDExt: Optional[str] = None,
+        openPMDInfix: Optional[str] = None,
+        timeAveraging: bool = True,
```
do we really want to default to time averaging?

ikbuibui replied:
Its the current behaviour in C++. Im not sure either if it is a sensible default. Time averaging applies when dump period is set to greater than 1. Then it averages the values in the binning by the dump period. The original reasoning behind making it enabled by default was so that values in bins can be understood and compared independent of the sump period.

BrianMarre replied:
Okay, we should keep the C++ and python defaults consistent, although I would propose to change both defaults in a follow up PR

BrianMarre commented on lib/python/picongpu/picmi/diagnostics/binning.py:
```diff
@@ -0,0 +1,191 @@
+        openPMDExt: Optional[str] = None,
+        openPMDInfix: Optional[str] = None,
+        timeAveraging: bool = True,
+        dumpPeriod: int = 1,
```
probably a dumb question what happens in the binning plugin if dumpPeriod < period ?

ikbuibui replied:
Dump period is defined as number of notifies after which a file is written out, so it is orthogonal to the period. A period of 5 and a dump period of 2 would mean a file is written every 10 steps, with data accumulated over 2 notifies.

BrianMarre replied:
Thanks for the clarification, this should be documented here to ensure that the user knows that dumpPeriod is in notfies, not simulation-steps as I assumed
